### PR TITLE
ProtocolV2: Payload recipients are matched on peerID + DID

### PR DESCRIPTION
**Pull request by gerardsn**
#1874 changed connection management from peerID to nodeDID+peerID. This change made it possible to have duplicate peerIDs in the connections list which hypothetically meant that a payload could end up at a different node than the one requesting it.

### CODE REVIEW

**Comment by gerardsn** [comment]
#1940 extends this to respond directly on the connection a message came from

File: network/network_integration_test.go [hunk 1]
```diff
@@ -741,6 +741,70 @@ func TestNetworkIntegration_PrivateTransaction(t *testing.T) {
...
+			mutex.Lock()
+			defer mutex.Unlock()
+			for _, receivedDoc := range receivedTransactions["node2"] {
+				if tx.Ref().Equals(receivedDoc.Ref()) {
+					return true, nil
+				}
+			}
+			return false, nil
+		}, time.Second)
+		arrivedAtEve := test.WaitForNoFail(t, func() (bool, error) {
```
Comment (reinkrul) [inline network/network_integration_test.go:793]: did you test the test (did it fail before you fixed the vulnerability)?

File: network/network_integration_test.go [hunk 2]
```diff
@@ -741,6 +741,70 @@ func TestNetworkIntegration_PrivateTransaction(t *testing.T) {
...
+			for _, receivedDoc := range receivedTransactions["node3"] {
+				if tx.Ref().Equals(receivedDoc.Ref()) {
+					return true, nil
+				}
+			}
+			return false, nil
+		}, time.Second)
+
+		// make sure eve does not have the payload
+		assert.False(t, arrivedAtEve)  // is true
```
Comment (reinkrul) [inline network/network_integration_test.go:805]: weird comments? (is true/ is false)

File: network/transport/v2/handlers.go [hunk 3]
```diff
@@ -133,24 +134,30 @@ func (p *protocol) handleTransactionPayloadQuery(ctx context.Context, peer trans
 		WithField(core.LogFieldTransactionRef, hash.FromSlice(msg.TransactionRef)).
 		Trace("Handling TransactionPayloadQuery")
 
+	// get peer's connection, authentication status is checked later on
+	connection := p.connectionList.Get(grpc.ByConnected(), grpc.ByNodeDID(peer.NodeDID), grpc.ByPeerID(peer.ID))
```
Comment (reinkrul) [inline network/transport/v2/handlers.go:138]: does this also work for unauthenticated connections? (bootstrapping?)

File: network/transport/v2/handlers.go [hunk 4]
```diff
@@ -133,24 +134,30 @@ func (p *protocol) handleTransactionPayloadQuery(ctx context.Context, peer trans
...
 			// Transaction not found
-			return p.send(peer, emptyResponse)
+			return connection.Send(p, &Envelope{Message: emptyResponse}, false)
 		}
 		return err
 	}
 	if len(tx.PAL()) > 0 {
 		// Private TX, verify connection
-		if !peer.Authenticated {
+		if !peer.Authenticated || !connection.Peer().Authenticated { // peer == connection.Peer()
```
Comment (reinkrul) [inline network/transport/v2/handlers.go:154]: if they're the same, why is the extrea check required?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: network/network_integration_test.go (hunk 1 above)
Comment (gerardsn) [reply network/network_integration_test.go:793]: first commit was test only, and yes it did fail

File: network/network_integration_test.go (hunk 2 above)
Comment (gerardsn) [reply network/network_integration_test.go:805]: removed. leftover from comment from when it was still failing

File: network/transport/v2/handlers.go (hunk 3 above)
Comment (gerardsn) [reply network/transport/v2/handlers.go:138]: yes, but public payloads are attached to their transactions so this is for private TXs only.
#1940 changes this to respond on the connection the query came from.

File: network/transport/v2/handlers.go (hunk 4 above)
Comment (gerardsn) [reply network/transport/v2/handlers.go:154]: being extra safe. removed in #1940

Comment (reinkrul) [reply network/transport/v2/handlers.go:154]: "being extra safe" sounds like unsure of program flow (or skills). But if it's removed in the next PR, ok.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**